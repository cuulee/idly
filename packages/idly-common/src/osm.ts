export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
