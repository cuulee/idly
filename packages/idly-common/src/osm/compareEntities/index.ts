export { compareNode } from './compareNode';
export { compareRelation } from './compareRelation';
export { compareTags } from './compareTags';
export { compareWay } from './compareWay';
