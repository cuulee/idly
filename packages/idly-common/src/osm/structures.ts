export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export interface Tags {
  readonly [key: string]: string;
}

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export interface FeatureProps {
  [key: string]: string;
}

export type NodeGeometry = OsmGeometry.POINT | OsmGeometry.VERTEX;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation',
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation',
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: ReadonlyArray<EntityId>;
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: ReadonlyArray<RelationMember>;
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: string;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

// id & ref are the same thing
// id exists for backward compatibility with iD
// we are using ref
export interface RelationMember {
  readonly id: string;
  readonly ref: string;
  readonly type?: string;
  readonly role?: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
