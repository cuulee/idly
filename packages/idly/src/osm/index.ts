// import { parse } from 'osm/network/parser';
export { nodeFactory } from 'osm/entities/node';
export { propertiesGen } from 'osm/others/properties';
export { wayFactory } from 'osm/entities/way';
export { graphFactory } from 'osm/history/graph';
