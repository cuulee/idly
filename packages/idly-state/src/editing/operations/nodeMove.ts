import { relationFactory, wayFactory } from 'idly-common/lib/osm/entityFactory';
import {
  Entity,
  LngLat,
  Node,
  Relation,
  Way,
} from 'idly-common/lib/osm/structures';
import { Table } from '../../table';
import { nodeUpdate } from '../helpers/nodeUpdate';
import { nodeUpdateParentWays } from '../helpers/nodeUpdateParentWays';
import { wayUpdateNodeId } from '../helpers/wayUpdateNodeId';
import { nodeUpdateParents } from './nodeUpdateParents';

export function nodeMove(
  prevNode: Node,
  newLoc: LngLat,
  parentWaysTable: Table<Way[]>,
  parentRelationsTable: Table<Relation[]>,
  idGen: (e: Entity) => string
): Entity[] {
  const newNode = nodeUpdate({ loc: newLoc, id: idGen(prevNode) }, prevNode);
  return [
    newNode,
    ...nodeUpdateParents(
      prevNode,
      newNode,
      parentWaysTable,
      parentRelationsTable,
      idGen
    ),
  ];
}
