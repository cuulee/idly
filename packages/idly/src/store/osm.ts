import { LngLatBounds } from 'mapbox-gl';
import { memoize } from 'ramda';
var SphericalMercator = require('@mapbox/sphericalmercator');

import { fetchTile } from '../network/osm';
import { cancelablePromise } from '../network/helper';
var mercator = new SphericalMercator({
  size: 256
});
const memFetchTile: (
  x: number,
  y: number,
  zoom: number
) => Promise<object> = memoize(fetchTile);

var prevProm: any = [];
export function fetchBbox(ltlng: LngLatBounds, zoom: number = 16) {
  prevProm.forEach((p: any) => p.cancel());
  const { minX, minY, maxX, maxY } = mercator.xyz(
    [ltlng.getWest(), ltlng.getSouth(), ltlng.getEast(), ltlng.getNorth()],
    Math.floor(zoom)
  );
  var zoom = Math.floor(zoom);
  var prom = [];
  for (var x = minX; x <= maxX; x++) {
    for (var y = minY; y <= maxY; y++) {
      prom.push(cancelablePromise(memFetchTile(x, y, zoom)));
    }
  }
  prevProm = prom.slice(0);
  return prom;
}
