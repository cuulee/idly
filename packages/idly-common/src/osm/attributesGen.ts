import { Attributes } from '../osm/structures';
const attr = {};

export function attributesGen(
  {
    visible,
    version,
    timestamp,
    changeset,
    uid,
    user
  }: {
    visible?: boolean;
    version?: string;
    timestamp?: string;
    changeset?: string;
    uid?: string;
    user?: string;
  } = {}
): Attributes {
  return {
    changeset,
    timestamp,
    uid,
    user,
    version,
    visible
  };
}
