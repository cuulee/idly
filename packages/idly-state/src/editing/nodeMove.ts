import { Entity, LngLat } from 'idly-common/lib/osm/structures';
import {
  entitiesMakeParentRelationsLookup,
  entityIdParentWays,
} from '../state/derivedTable/helpers';
import { genNextId } from '../stateHelpers/genNextId';
import { getLatestId } from '../stateHelpers/getLatestId';
import { OsmState } from '../type';
import { nodeMoveOp } from './operations/nodeMove';

export function nodeMove(
  state: OsmState,
  nodeId: string,
  loc: LngLat
): Entity[] {
  const id = getLatestId(state.log, nodeId);
  const derived = state.derivedTable.get(id);

  if (!derived) {
    throw new Error('couldnt create entry as no derived values found' + id);
  }

  const relationsLookup = entitiesMakeParentRelationsLookup(
    [id, ...derived.parentWays],
    state.derivedTable
  );

  return nodeMoveOp({
    idGen: (e: Entity) => genNextId(state, e.id),
    newLoc: loc,
    parentRelationsLookup: relationsLookup,
    parentWays: entityIdParentWays(id, state.derivedTable),
    prevNode: derived.entity as any,
  });
}
