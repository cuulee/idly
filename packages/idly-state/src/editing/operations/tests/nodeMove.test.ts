import {
  nodeFactory,
  relationFactory,
  wayFactory,
} from 'idly-common/lib/osm/entityFactory';
import { simpleIdIncr } from '../../helpers/tests/util';
import { nodeMove } from '../nodeMove';
import { Relation } from 'idly-common/lib/osm/structures';

const mapFromObj = (o: any): Map<string, any> =>
  Object.keys(o).reduce((prev, k) => {
    prev.set(k + '', o[k]);
    return prev;
  }, new Map());

const n1 = nodeFactory({
  id: 'n1',
});
const n2 = nodeFactory({
  id: 'n2',
});
const n3 = nodeFactory({
  id: 'n3',
});
const n4 = nodeFactory({
  id: 'n4',
});
const n5 = nodeFactory({
  id: 'n5',
});

const n6 = nodeFactory({
  id: 'n6',
});

const n7 = nodeFactory({
  id: 'n7',
});

const w1 = wayFactory({
  id: 'w1',
  nodes: ['n1', 'n2'],
});

const w2 = wayFactory({
  id: 'w2',
  nodes: ['n1', 'n3'],
});

const w3 = wayFactory({
  id: 'w3',
  nodes: ['n3', 'n4', 'n5'],
});

const r1 = relationFactory({
  id: 'r1',
  members: [{ id: 'n4', ref: 'n4' }, { id: 'w2', ref: 'w2' }],
});

const r2 = relationFactory({
  id: 'r2',
  members: [
    { id: 'w1', ref: 'w1' },
    { id: 'w3', ref: 'w3' },
    { id: 'n4', ref: 'n4' },
    { id: 'n6', ref: 'n6' },
  ],
});

test('move a node with less dependencies', () => {
  const prevNode = n1;
  const newEntities = nodeMove(
    prevNode,
    { lat: 5, lon: 10 },
    mapFromObj({
      n1: [w1, w2],
    }),
    mapFromObj({
      n1: [],
      w1: [r2],
      w2: [r1],
    }),
    simpleIdIncr
  );

  expect(newEntities.map(r => r.id)).toEqual([
    'n1#0',
    'w1#0',
    'w2#0',
    'r2#0',
    'r1#0',
  ]);

  expect(newEntities.find(r => r.id === 'w1#0')).toMatchSnapshot();
  expect(newEntities.find(r => r.id === 'w2#0')).toMatchSnapshot();
  expect(newEntities.find(r => r.id === 'r2#0')).toMatchSnapshot();
});

test('move a node where the node and parentWay share a common relation', () => {
  const prevNode = n4;
  const newEntities = nodeMove(
    prevNode,
    { lat: 5, lon: 10 },
    mapFromObj({
      n4: [w3],
    }),
    mapFromObj({
      garbage: [],
      n4: [r1, r2],
      w3: [r2],
    }),
    simpleIdIncr
  );

  expect(newEntities.map(r => r.id)).toEqual(['n4#0', 'w3#0', 'r1#0', 'r2#0']);
  expect(
    (newEntities.find(r => r.id === 'r2#0') as Relation).members.map(r => r.id)
  ).toEqual(['w1', 'w3#0', 'n4#0', 'n6']);
  expect(
    (newEntities.find(r => r.id === 'r1#0') as Relation).members.map(r => r.id)
  ).toEqual(['n4#0', 'w2']);
});
