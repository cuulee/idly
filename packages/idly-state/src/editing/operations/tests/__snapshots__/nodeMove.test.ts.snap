// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`move a node with less dependencies 1`] = `
Object {
  "attributes": Object {
    "changeset": undefined,
    "timestamp": undefined,
    "uid": undefined,
    "user": undefined,
    "version": undefined,
    "visible": undefined,
  },
  "id": "w1#0",
  "nodes": Array [
    "n1#0",
    "n2",
  ],
  "tags": Object {},
  "type": "way",
}
`;

exports[`move a node with less dependencies 2`] = `
Object {
  "attributes": Object {
    "changeset": undefined,
    "timestamp": undefined,
    "uid": undefined,
    "user": undefined,
    "version": undefined,
    "visible": undefined,
  },
  "id": "w2#0",
  "nodes": Array [
    "n1#0",
    "n3",
  ],
  "tags": Object {},
  "type": "way",
}
`;

exports[`move a node with less dependencies 3`] = `
Object {
  "attributes": Object {
    "changeset": undefined,
    "timestamp": undefined,
    "uid": undefined,
    "user": undefined,
    "version": undefined,
    "visible": undefined,
  },
  "id": "r2#0",
  "members": Array [
    Object {
      "id": "w1#0",
      "ref": "w1#0",
    },
    Object {
      "id": "w3",
      "ref": "w3",
    },
    Object {
      "id": "n4",
      "ref": "n4",
    },
    Object {
      "id": "n6",
      "ref": "n6",
    },
  ],
  "tags": Object {},
  "type": "relation",
}
`;
